plugin/utils/shell.py

```diff
@@ -113,145 +124,194 @@ def generate_rig_edit():
+        if (not vectorisclose(p_bone.location,Vector((0,0,0)),0.0001) or not vectorisclose(p_bone.scale,Vector((1,1,1)),0.0001) or not vectorisclose(Vector(p_bone.rotation_quaternion),Vector((1,0,0,0)),0.0001)) and p_bone.name.startswith("NODE_"):
+            #Ignore posed NODE bones and proceed to the next, their offsets can be applied directly.
             editnumber = editnumber + 1
+            logging.info(f"rig edit number {editnumber}")
+            logging.info(f"NODE with offsets detected. Applying offsets directly.")
+            continue
+        
+        #We append any remaining bones that have been posed.
+        if (not vectorisclose(p_bone.location,Vector((0,0,0)),0.0001) or not vectorisclose(p_bone.scale,Vector((1,1,1)),0.0001) or not vectorisclose(Vector(p_bone.rotation_quaternion),Vector((1,0,0,0)),0.0001)):
+            #Append the bones to the list of posed bones.
+            posedbones_list.append(p_bone)
+    
+    
+    #Iterate over every posed bone.
+    for p_bone in posedbones_list:
+        #Update bone locations
+        bpy.context.view_layer.update()
+            
+        #Debug info
+        logging.info(f"rig edit number {editnumber+1}")
+        logging.info(f"name: {p_bone.name}")
+        logging.info(f"parent: {p_bone.parent.name}")
+        
+        #Get pose mode position of bonebase
+        bonebase_pos_pose = Vector(p_bone.matrix.translation)
+        
+        #Reset position
+        p_bone.location = Vector((0, 0, 0))
+        #Update location after moving.
+        bpy.context.view_layer.update()
+        
+        #Get rest mode position of bonebase
+        bonebase_pos_rest = Vector(p_bone.matrix.translation)
+        
+        #Calculate delta offset
+        bonebase_pos_delt = Vector(bonebase_pos_pose - bonebase_pos_rest)
+        
+        #Set back to original position
+        p_bone.matrix.translation = Vector(bonebase_pos_pose)
+        
+        #Log locations
+        logging.info(f"bonebase rest position: {bonebase_pos_rest}")
+        logging.info(f"bonebase pose position: {bonebase_pos_pose}")
+        logging.info(f"bonebase delt position: {bonebase_pos_delt}")
+
+        
+        #Creating the nodes            
+        #We switch to edit mode to use edit bones, as they do not exist outside of edit mode.
+        #Switch to edit mode
+        bpy.ops.object.mode_set(mode='EDIT')
+        
+        #Get edit bones
+        bonebase = armature.data.edit_bones.get(p_bone.name)
+        boneparent = armature.data.edit_bones.get(p_bone.parent.name)
+        
+        #Creating the node bone
+        bonenode = armature.data.edit_bones.new(f"NODE_{p_bone.name}")
+        #NOTE: Maybe here we set the bonenode to boneparent if it starts with NODE_ already
+        
+        #Copy parent length to node
+        bonenode.length = boneparent.length
+        #Copy parent matrix to node
+        bonenode.matrix = boneparent.matrix.copy()
```

Comment:
Setting the matrix is known not to be reliable for bone roll. Better copy head, tail, roll